text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,518 @@
+   the large size of `aws-cdk-lib`.
+6. The developer experience of using `aws-cdk-lib` in an IDE might be slower in
+   some cases. 
+7. The CDK project is currently blocked from including more ‘large
+   dependencies’, such as multiple version of kubectl, multiple version of the
+   AWS CLI, and dedicated versions of the AWS SDK for use in custom resources.
+   This is a decision we’ve made to avoid increasing the size of the package
+   further until a more scalable solution for these kinds of dependencies is
+   implemented.
+
+Most of these above issues have workarounds. For running CDK apps in Lambda, you
+can use a Docker image for your Lambda code, and the limit becomes 10GB. For
+package manager size limits, we can request
+[increases](https://github.com/pypa/pypi-support/issues/1642), and those have
+historically been approved. However, publishing a large package is a bad
+practice in the open-source software ecosystem. Even if a customer has no
+technical limitations for adding a ~240 MB dependency to their package, it is a
+red flag to them when considering using `aws-cdk-lib`. We should be
+customer-obsessed, and solve the problem of aws-cdk-lib being a huge dependency,
+before customers start complaining about it.
+
+### Why is the package so large?
+
+There are some large assets contributing to the size of aws-cdk-lib, like .jsii
+files, and zip files of dependencies used in custom resources. However, the
+library also contains a huge volume of source code now that `aws-cdk-lib` combines
+231 (and increasing) CDK modules. The table below breaks down the percentage
+that each category of files contributes to the size of aws-cdk-lib. 
+
+Note: The percentages do not add up to 100. Some files, e.g. various .json
+files, are excluded and contribute very little to the size. 
+
+| Category                       | Percentage of aws-cdk-lib size |
+| ------------------------------ | ------------------------------ |
+| .jsii files                    | 41.64%                         |
+| Source Maps (.js.map)          | 21.93%                         |
+| Lambda Layer zipfiles          | 15.48%                         |
+| Javascript code (.js)          | 9.17%                          |
+| Type Declaration files (.d.ts) | 9.15%                          |
+| README, etc (.md)              | 0.76%                          |
+| bundled npm dependencies       | 0.69%                          |
+| .ts-fixture                    | 0.37%                          |
+| .jsiirc                        | 0.36%                          |
+
+
+
+### Why should we _not_ do this?
+
+The main reason to not do this is that it will inevitably break up the
+monolothic `aws-cdk-lib` package in some manner, and we will lose some of the
+wonderful simplicity of releasing and using a single package. But, we can't keep
+everything in one place, and reduce the size enough to have a significant
+positive impact.
```

Comment:
Yeah, these are good points. The current biggest contributor of size is not the Lambda Layers, they are just the biggest potential contributor. I called this out in the execution plan, but would be good to add it here. Here we should also call out the approximate size of monolithic aws-cdk-lib with all of the extra Lambda Layers. With .jsii compression and removing source maps it would still be nearly 200 MB.

Reply:
It feels like the argument of this RFC hinges on this future size that includes additional lambda layers. I think it's a bit confusing to have both current and future estimates here, and instead we should benchmark _only_ on the future estimate. Then all the sections will have the same reference point and we don't have to switch context between comapring the impact of .jsii file compression to the current size and lambda layers to the future size.

I'm open to the response "too much effort / not needed" but that's just my $0.02.

Reply:
Hmm, I hadn't thought of it that way. Good idea. Let me switch to framing it that way, and see how it looks. I want to avoid including within the scope of work actually adding the future lambda layers. This project should just set up the mechanism to add them.